testonly/hammer/contents.go

```diff
@@ -0,0 +1,200 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hammer
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian"
+)
+
+// mapContents is a complete copy of the map's contents at a particular revision.
+type mapContents struct {
+	data map[mapKey]string
+	rev  int64
+}
+type mapKey [sha256.Size]byte
+
+func (m *mapContents) empty() bool {
+	return len(m.data) == 0
+}
+
+// pickKey randomly select a key that already exists in a given copy of the
+// map's contents. Assumes that the selected copy is non-empty.
+func (m *mapContents) pickKey(prng rand.Source) []byte {
+	if m.empty() {
+		panic("internal error: can't pick a key, map data is empty!")
+	}
+
+	choice := intN(prng, len(m.data))
+	// Need sorted keys for reproduceability.
+	keys := make([]mapKey, 0)
+	for k := range m.data {
+		keys = append(keys, k)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return bytes.Compare(keys[i][:], keys[j][:]) == -1
+	})
+	for _, k := range keys {
+		if choice == 0 {
+			return k[:]
+		}
+		choice--
+	}
+	panic("internal error: inconsistent count of map contents")
+}
+
+// checkContents compares information returned from the Map against a local copy
+// of the map's contents.
+func (m *mapContents) checkContents(leafInclusions []*trillian.MapLeafInclusion) error {
+	for _, inc := range leafInclusions {
+		leaf := inc.Leaf
+		var key mapKey
+		copy(key[:], leaf.Index)
+		value, ok := m.data[key]
+		if ok {
+			if string(leaf.LeafValue) != value {
+				return fmt.Errorf("got leaf[%v].LeafValue=%q, want %q", key, leaf.LeafValue, value)
+			}
+			if want := "extra-" + value; string(leaf.ExtraData) != want {
+				return fmt.Errorf("got leaf[%v].ExtraData=%q, want %q", key, leaf.ExtraData, want)
+			}
+		} else {
+			if len(leaf.LeafValue) > 0 {
+				return fmt.Errorf("got leaf[%v].LeafValue=%q, want not-present", key, leaf.LeafValue)
+			}
+		}
+	}
+	return nil
+}
+
+// updatedWith returns a new mapContents object that has been updated to include the
+// given leaves and revision.  A nil receiver object is allowed.
+func (m *mapContents) updatedWith(rev uint64, leaves []*trillian.MapLeaf) *mapContents {
+	// Start from previous map contents
+	result := mapContents{rev: int64(rev), data: make(map[mapKey]string)}
+	if m != nil {
+		for k, v := range m.data {
+			result.data[k] = v
+		}
+	}
+	// Update with given leaves
+	for _, leaf := range leaves {
+		var k mapKey
+		copy(k[:], leaf.Index)
+		if leaf.LeafValue != nil {
+			result.data[k] = string(leaf.LeafValue)
+		} else {
+			delete(result.data, k)
+		}
+	}
+
+	return &result
+}
+
+// How many copies of map contents to hold on to.
+const copyCount = 10
+
+type previousMapContents struct {
```

Comment:
`multipleMapContents` might be clearer

Reply:
I went for `versionedMapContents` to try to make it clear it's the same map...